src/main/java/bookbob/functions/FileHandler.java

```diff
@@ -150,8 +152,65 @@ public static void retrieveData(Records records){
             logger.log(Level.WARNING, "File not found", e);
             throw new RuntimeException(e);
         }
+    }*/
+
+    //@@author kaboomzxc
+    public static void retrieveData(Records records) {
+        try {
+            File file = new File(filePath);
+            Scanner reader = new Scanner(file);
+
+            while (reader.hasNextLine()) {
+                String line = reader.nextLine();
+                String[] data = line.split("\\|");
+
+                if (data.length < 9) {
+                    continue;
+                }
+
+                // Parse basic patient information
+                String name = data[0].substring(6).trim();
+                String nric = data[1].substring(6).trim();
+                String phoneNumber = data[2].substring(15).trim();
+                String dateOfBirth = data[3].substring(16).trim();
+                String homeAddress = data[4].substring(15).trim();
+                String sex = data[6].substring(5).trim();
+                ArrayList<String> allergies = parseList(data[5].substring(9).trim());
+                ArrayList<String> medicalHistories = parseList(data[7].substring(17).trim());
+
+                // Parse visits
+                ArrayList<Visit> visits = new ArrayList<>();
+                String visitsString = data[8].trim();
+
+                // Extract the content between the outer-most brackets
+                int firstBracket = visitsString.indexOf("[");
+                int lastBracket = visitsString.lastIndexOf("]");
+
+                if (firstBracket != -1 && lastBracket != -1) {
+                    visitsString = visitsString.substring(firstBracket + 1, lastBracket);
+
+                    // Split into individual visits by looking for date pattern
+                    String[] visitParts = visitsString.split("(?<=\\]),\\s*(?=\\d{2}-\\d{2}-\\d{4})");
+
+                    for (String visitPart : visitParts) {
+                        Visit visit = parseVisitInputString("[" + visitPart + "]");
+                        if (visit != null) {
+                            visits.add(visit);
+                        }
+                    }
+                }
```

Comment:
Perhaps can consider avoiding deeply nested codes (arrowhead codes) for better code quality. I also noticed this in other parts of the code such as FileHandler.java